srml/contract/src/lib.rs

```diff
@@ -135,14 +136,14 @@ where
 }
 
 /// The default dispatch fee computor computes the fee in the same way that
-/// implementation of `MakePayment` for balances module does.
+/// implementation of `ChargeBytesFee` for fees module does.
 pub struct DefaultDispatchFeeComputor<T: Trait>(PhantomData<T>);
 impl<T: Trait> ComputeDispatchFee<T::Call, T::Balance> for DefaultDispatchFeeComputor<T> {
 	fn compute_dispatch_fee(call: &T::Call) -> T::Balance {
 		let encoded_len = parity_codec::Encode::encode(&call).len();
-		let base_fee = <balances::Module<T>>::transaction_base_fee();
-		let byte_fee = <balances::Module<T>>::transaction_byte_fee();
-		base_fee + byte_fee * <T::Balance as As<u64>>::sa(encoded_len as u64)
+		let base_fee = <fees::Module<T>>::transaction_base_fee();
+		let byte_fee = <fees::Module<T>>::transaction_byte_fee();
+		<T::Balance as As<u64>>::sa(base_fee.as_() + byte_fee.as_() * encoded_len as u64)
```

Comment:
Because contract want to charge fee in gas instead of balance it reproduce `charge_base_bytes_fee` here in its `DefaultDispatchFeeComputer`
It seems it would preferable to call a function like `compute_base_bytes_fee(encoded_len)` here don't you think so ?

Maybe we could add this to ChargeByteFee
```rust
/// Charge bytes fee trait
pub trait ChargeBytesFee<AccountId> {
    type Amount;
    fn charge_base_bytes_fee(transactor: &AccountId, encoded_len: usize) -> Result<(), &'static str>;
    fn compute_base_bytes_fee(encoded_len: usize) -> Self::Amount; // But then it needs the type to return it.
}
```

But so it adds Amount accociated type and complexify the trait for a maybe very specific use case.
I'm not really in favor of adding this to the trait.

I would approve as is but this DefaultDispatchFeeComputor in contracts doesn't seems perfect

Reply:
I also came up with the idea of adding a `compute_base_bytes_fee` function to the trait, while replacing balances stuff here, but then realised this might be a overkill. And since `Balance` type is used heavily in contract module, doing this change now still won't make it perfect, as we need to translate between `Amount` and `Balance` types. Ideally fee charging related codes including gas fee should use `Amount` type, to decouple from `balances` module.

I'll work on another task trying to move transfer/creation fee from `balances` module to `fees`, and decouple `balances` with `contract` module. Refactoring `DefaultDispatchFeeComputor` might fit in that task more.

Reply:
It seems @thiolliere will be working to update the `contract` module to remove balances from it, I think he might be able to make the update he proposes at that time if it is needed then. At the moment, it looks like this PR is at a pretty good spot, and I think we shouldn't block on this idea to get it merged in

Reply:
@shaopengw Yes I thought about that when working on https://github.com/paritytech/substrate/pull/1641
I am currently working on something else but still have some ideas on it. Feel free to work on it I struggle to find a way to do it, also ping me on that.

My main thoughts are: should we make the overlay as a trait implemented by balances, or make a low level trait that has enough to be able to implement this overlay ?
and some ideas are in this comment: https://github.com/paritytech/substrate/pull/1641#issuecomment-460676981

Reply:
@thiolliere Thanks for sharing. I'll open an issue to draft the design and would pm you, we could discuss details there :]